### PR TITLE
persist oom files on cgroup v2

### PR DESCRIPTION
**Pull request by haircommander**
fixes: https://github.com/containers/conmon/pull/97#issuecomment-565065081

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

**Comment by haircommander** [comment]
PTAL @giuseppe @mrunalp @saschagrunert @rhatdan

File: src/conmon.c [hunk 1]
```diff
@@ -618,13 +626,8 @@ static gboolean check_cgroup2_oom()
...
 
 		if (counter != last_counter) {
-			_cleanup_close_ int oom_fd = -1;
-
-			ninfo("OOM received");
-			oom_fd = open("oom", O_CREAT, 0666);
-			if (oom_fd < 0) {
-				nwarn("Failed to write oom file");
-			} else {
+			int oom_fd = write_oom_files();
```
Comment (mrunalp) [inline src/conmon.c]: Are we cleaning up oom_fd?

File: src/conmon.c [hunk 2]
```diff
@@ -547,6 +547,26 @@ static bool has_v1_oom()
 	return false;
 }
 
+/* write the appropriate files to tell the caller there was an oom event
+ * this can be used for v1 and v2 OOMS */
+static int write_oom_files() {
+	_cleanup_close_ int oom_fd = -1;
```
Comment (mrunalp) [inline src/conmon.c]: Can we check if this works with us returning oom_fd here?

Comment (haircommander) [reply src/conmon.c]: ope, it did not. Fixed

File: src/conmon.c (hunk 1 above)
Comment (haircommander) [reply src/conmon.c]: now we are

**Comment by haircommander** [comment]
> To me it looks good for what I can tell. I think we should find a way to add some testing around this.

agreed, we need unit/e2e tests bad

**Comment by mrunalp** [comment]
LGTM. @giuseppe ptal.

File: src/conmon.c [hunk 3]
```diff
@@ -547,6 +547,28 @@ static bool has_v1_oom()
 	return false;
 }
 
+/* write the appropriate files to tell the caller there was an oom event
+ * this can be used for v1 and v2 OOMS
+ * Note: the caller is responsible for closing oom_fd
+ */
+static int write_oom_files() {
+	int oom_fd = -1;
```
Comment (giuseppe) [inline src/conmon.c]: could we have the `_cleanup_close_` here?

File: src/conmon.c [hunk 4]
```diff
@@ -547,6 +547,28 @@ static bool has_v1_oom()
...
+		ctr_oom_fd = open(ctr_oom_file_path, O_CREAT, 0666);
+		if (ctr_oom_fd < 0) {
+			nwarn("Failed to write oom file");
+		}
+	}
+	oom_fd = open("oom", O_CREAT, 0666);
+	if (oom_fd < 0) {
+		nwarn("Failed to write oom file");
+	}
+	return oom_fd;
```
Comment (giuseppe) [inline src/conmon.c]: it could be turned into:
```suggestion
	return oom_fd >= 0 ? 0 : -1;
```

File: src/conmon.c [hunk 5]
```diff
@@ -560,22 +582,10 @@ static gboolean oom_cb_cgroup_v1(int fd, GIOCondition condition, G_GNUC_UNUSED g
...
-			}
-			if (opt_persist_path) {
-				_cleanup_close_ int ctr_oom_fd = -1;
-				_cleanup_free_ char *ctr_oom_file_path = g_build_filename(opt_persist_path, "oom", NULL);
-				ctr_oom_fd = open(ctr_oom_file_path, O_CREAT, 0666);
-				if (ctr_oom_fd < 0) {
-					nwarn("Failed to write oom file");
-				}
-			}
+			_cleanup_close_ int oomfd = write_oom_files();
```
Comment (giuseppe) [inline src/conmon.c]: then here you don't need to declare `oomfd`

File: src/conmon.c [hunk 6]
```diff
@@ -618,13 +628,8 @@ static gboolean check_cgroup2_oom()
...
 		if (counter != last_counter) {
-			_cleanup_close_ int oom_fd = -1;
-
-			ninfo("OOM received");
-			oom_fd = open("oom", O_CREAT, 0666);
-			if (oom_fd < 0) {
-				nwarn("Failed to write oom file");
-			} else {
+			_cleanup_close_ int oom_fd = write_oom_files();
+			if (oom_fd >= 0) {
```
Comment (giuseppe) [inline src/conmon.c]: and here as well you wouldn't need to declare `oom_fd`:

```
			if (write_oom_files() == 0) {
```

**Comment by haircommander** [comment]
thanks @giuseppe, updated as suggested

**Comment by giuseppe** [comment]
Thanks, LGTM

**Pull request merged**